waspc/src/Wasp/Project/Analyze.hs

```diff
@@ -69,8 +84,107 @@ analyzeWaspProject waspDir options = do
+  IO (Either [CompileError] (Path' Abs (File CompiledWaspJsFile)))
+compileWaspTsFile waspProjectDir tsconfigNodeFileInWaspProjectDir waspFilePath = do
+  chan <- newChan
+  (_, tscExitCode) <-
+    concurrently
+      (readJobMessagesAndPrintThemPrefixed chan)
+      ( runNodeCommandAsJob
+          waspProjectDir
+          "npx"
+          [ "tsc",
+            "-p",
+            toFilePath (waspProjectDir </> tsconfigNodeFileInWaspProjectDir),
+            "--noEmit",
+            "false",
+            "--outDir",
+            toFilePath outDir
+          ]
+          J.Wasp
+          chan
+      )
+  case tscExitCode of
+    ExitFailure _status -> return $ Left ["Got TypeScript compiler errors for " ++ toFilePath waspFilePath ++ "."]
+    ExitSuccess -> return $ Right absCompiledWaspJsFile
+  where
+    outDir = waspProjectDir </> dotWaspDirInWaspProjectDir
+    absCompiledWaspJsFile = outDir </> compiledWaspJsFileInDotWaspDir
+    compiledWaspJsFileInDotWaspDir = SP.castFile $ case replaceRelExtension (basename waspFilePath) ".mjs" of
+      Just path -> path
+      Nothing -> error $ "Couldn't calculate the compiled JS file path for " ++ toFilePath waspFilePath ++ "."
+
+executeMainWaspJsFile ::
+  Path' Abs (Dir WaspProjectDir) ->
+  Psl.Schema.Schema ->
+  Path' Abs (File CompiledWaspJsFile) ->
+  IO (Either [CompileError] (Path' Abs (File DeclsJsonFile)))
+executeMainWaspJsFile waspProjectDir prismaSchemaAst absCompiledMainWaspJsFile = do
+  chan <- newChan
+  (_, runExitCode) <- do
+    concurrently
+      (readJobMessagesAndPrintThemPrefixed chan)
+      ( runNodeCommandAsJob
+          waspProjectDir
+          "npx"
+          -- TODO: Figure out how to keep running instructions in a single
+          -- place (e.g., this is string the same as the package name, but it's
+          -- repeated in two places).
+          -- Before this, I had the entrypoint file hardcoded, which was bad
+          -- too: waspProjectDir </> [relfile|node_modules/wasp-config/dist/run.js|]
+          [ "wasp-config",
+            SP.fromAbsFile absCompiledMainWaspJsFile,
+            SP.fromAbsFile absDeclsOutputFile,
+            encodeToString allowedEntityNames
+          ]
+          J.Wasp
+          chan
+      )
+  case runExitCode of
+    ExitFailure _status -> return $ Left ["Error while running the compiled *.wasp.mts file."]
+    ExitSuccess -> return $ Right absDeclsOutputFile
+  where
+    absDeclsOutputFile = waspProjectDir </> dotWaspDirInWaspProjectDir </> [relfile|decls.json|]
+    allowedEntityNames = Psl.Schema.getModelNames prismaSchemaAst
+
+readDeclsFromJsonFile :: Path' Abs (File DeclsJsonFile) -> IO (Either [CompileError] [AS.Decl])
+readDeclsFromJsonFile declsJsonFile = do
+  declsBytestring <- IOUtil.readFileBytes declsJsonFile
+  case Aeson.eitherDecode declsBytestring of
+    Right value -> return $ Right value
+    Left err -> return $ Left ["Error while parsing the declarations from JSON: " ++ err]
```

Comment:
Extract `return` and use `fromMaybe` or that `fromJustOrElse` that I mention above.

Reply:
This is an Either so I used `left`.